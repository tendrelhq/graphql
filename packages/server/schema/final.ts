import { resolvers, typeDefs } from "@/schema";
import { makeExecutableSchema, mergeSchemas } from "@graphql-tools/schema";
import type { GraphQLScalarType } from "graphql";
import { type GqlUrl, register } from "./system/scalars/url";
import { getSchema } from "./v1.schema";

const legacySchema = makeExecutableSchema({ typeDefs, resolvers });
const gratsSchema = getSchema();

export const schema = mergeSchemas({ schemas: [legacySchema, gratsSchema] });

// Custom scalars. See https://grats.capt.dev/docs/docblock-tags/scalars/#serialization-and-parsing-of-custom-scalars.
register(schema.getType("URL") as GraphQLScalarType<GqlUrl, string>);
