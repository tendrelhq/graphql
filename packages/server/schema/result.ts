/** @gqlType */
export type Diagnostic = {
  __typename: "Diagnostic";

  /** @gqlField */
  code: DiagnosticKind;

  /** @gqlField */
  message?: string | null;
};

/** @gqlEnum */
export enum DiagnosticKind {
  // This feels wrong. Presumably this implies a no-op.
  candidate_change_discarded = "candidate_change_discarded",
  // This maybe also feels wrong.
  candidate_choice_unavailable = "candidate_choice_unavailable",
  /**
   * Indicates that an operation expected a template type to be provided but
   * received an instance type.
   */
  expected_template_got_instance = "expected_template_got_instance",
  /**
   * Indicates that an operation expected an instance type to be provided but
   * received a template type.
   */
  expected_instance_got_template = "expected_instance_got_template",
  /*
   * Diagnostics of this kind indicate that a feature is being used that is not
   * yet available for public consumption.
   */
  feature_not_available = "feature_not_available",
  /**
   * Some operations accept an optional hash. This is misleading. You should
   * _always_ pass a hash for operations that accept them.
   */
  hash_is_required = "hash_is_required",
  /**
   * Diagnostics of this kind indicates that the requested operation is no longer
   * a valid operation due to a state change that has not yet been observed by
   * the client. Typically this is due to data staleness but may also occur for
   * the _loser_ of a race under concurrency.
   *
   * Hashes are opaque. Clients should not attempt to derive any meaning from them.
   */
  hash_mismatch_precludes_operation = "hash_mismatch_precludes_operation",
  /**
   * Indicates that an operation received a type that it is not allowed to
   * operate on.
   */
  invalid_type = "invalid_type",
  /**
   * When you operate on a StateMachine<T>, there must obviously be a state
   * machine to operate *on*. This diagnostic is returned when no such state
   * machine exists.
   */
  no_associated_fsm = "no_associated_fsm",
}
