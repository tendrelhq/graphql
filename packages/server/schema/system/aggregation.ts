/** @gqlType */
export type Aggregate = {
  /**
   * The group, or bucket, that uniquely identifies this aggregate.
   * For example, this will be one of the `overType`s passed to `chainAgg`.
   *
   * @gqlField
   */
  group: string;

  /**
   * The computed aggregate value.
   *
   * Currently, this will always be a string value representing a duration in
   * seconds, e.g. "360" -> 360 seconds. `null` will be returned when no such
   * aggregate can be computed, e.g. "time in planned downtime" when no "planned
   * downtime" events exist. Note also that `null` will be returned if a
   * duration cannot be computed, e.g. because there is no end date.
   *
   * @gqlField
   */
  value: string | null;
};
