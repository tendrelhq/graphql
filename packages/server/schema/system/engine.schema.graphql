"""
State characterizes something, most commonly an entity, as existing in a
particular state. For example, an entity could be in a "pending" state.
"""
interface State {
  tag: String
}

type DebugInfo {
  createdAt: String!
  # createdBy: Identity
  updatedAt: String!
  updatedBy: Identity
}

type ComponentConnection {
  edges: [ComponentEdge!]!
  pageInfo: PageInfo!
  totalCount: Int!
}

type ComponentEdge {
  cursor: String!
  debug: DebugInfo!
  node: Component!
}

interface Referenceable implements Component {
  id: ID!
}

#
# Sort/filter/search primitives
#
"""
Defines the sort order for queries that support edge ordering.
Ordering depends on the scalar type:
  - Boolean scalars are sorted false < true
  - Int scalars are sorted numerically
  - String scalars are sorted lexicographically
"""
enum SortOrder {
  asc
  desc
}
