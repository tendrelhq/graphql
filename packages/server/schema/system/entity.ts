import { constructHeadersFromArgs, extractPageInfo } from "@/api";
import { getAccessToken, setCurrentIdentity } from "@/auth";
import { base_url } from "@/config";
import { sql } from "@/datasources/postgres";
import { assert, assertNonNull, assertUnderlyingType } from "@/util";
import { GraphQLError } from "graphql";
import type { ID, Int } from "grats";
import { decodeGlobalId, encodeGlobalId } from ".";
import type { Context } from "../types";
import { type Field, type FieldInput, field$fragment } from "./component";
import type { DisplayName } from "./component/name";
import { Task } from "./component/task";
import type { Connection, Edge } from "./pagination";

//-----------------------------------------------------------------------------
//
// Entity Instances
//
// Note that currently entity instances map directly to `entity.entityinstance`s
// by way of the Entity REST API.
//
//-----------------------------------------------------------------------------

/**
 * Entities represent distinct objects in the system. They can be physical
 * objects, like Locations, Resources and Workers, or logical ones, like
 * "Scan Codes".
 *
 * @gqlType
 */
export interface EntityInstance {
  readonly _type: string;
  readonly _id: string;

  /** @gqlField */
  id: ID;
}

/** @gqlField name */
export async function entityInstanceName(
  entity: EntityInstance,
  ctx: Context,
): Promise<DisplayName> {
  return await ctx.orm.displayName.load(entity.id);
}

/**
 * @gqlQueryField
 */
export async function instances(
  ctx: Context,
  args: {
    first?: Int | null;
    after?: string | null;
    /** TEMPORARY: this should be the customer/organization uuid. */
    owner: ID;
    /** Instances with the given parent (instance). */
    parent?: ID[] | null;
  },
): Promise<Connection<EntityInstance>> {
  const token = await getAccessToken(ctx.auth.userId)
    .then(r => r.json())
    .then(r => r.access_token);

  const headers = constructHeadersFromArgs(args, { count: "exact" });
  headers.set("Authorization", `Bearer ${token}`);

  const { type: ownerType, id: ownerId } = decodeGlobalId(args.owner);
  assert(ownerType === "organization", "`owner` must point to a customer!");

  // FIXME: Should not be required?
  const [{ owner }] = await sql`
    select entityinstanceuuid as owner
    from entity.entityinstance
    where entityinstanceoriginaluuid = ${ownerId}
  `;

  const q = new URLSearchParams({
    select: "id",
    owner: `eq.${owner}`,
    order: "_order.asc",
    _deleted: "eq.false",
  });
  if (args.parent?.length) {
    // FIXME: This is a weird one in Keller's entity model. Every entityinstance
    // has a parent, template and type. I'm not entirely sure why these three
    // exist. From looking through what Keller has done so far, this seems to be
    // the case, e.g. for a custag:
    // - template and type point at the "Customer Tag" template and instance, respectively
    // - parent points at the "Reason Code" instance
    q.append("parent", `in.(${args.parent.join(",")})`);
  }

  const res = await fetch(
    new URL(`/api/v1/entity_instance?${q.toString()}`, base_url),
    {
      method: "GET",
      headers: headers,
    },
  );

  if (!res.ok) {
    // TODO: Better error handling when interacting with the Entity API.
    const body = await res.json();
    console.error(body);
    throw new GraphQLError("Internal Server Error");
  }

  const rows: { id: string }[] = await res.json();
  return {
    edges: rows.map(row => ({
      cursor: "", // ignored but technically required by the connection spec
      node: {
        _id: row.id,
        _type: "entity_instance",
        id: encodeGlobalId({
          id: row.id,
          type: "entity_instance",
        }),
      },
    })),
    ...extractPageInfo(res),
  };
}

/**
 * Lift an EntityInstance to a set of Fields where the given EntityInstance
 * identifies as the Field's ValueType.
 *
 * @gqlField
 */
export async function asFieldTemplateValueType(
  edge: Edge<EntityInstance>,
): Promise<Connection<Field>> {
  // This is sort of half baked at the moment since we are only using this for
  // Runtime "reason codes". Additionally, work instances and templates are not
  // yet covered by the entity model.

  // The EntityInstance we have here is really a custag. We need to:
  // 1. Find the custag
  // 2. Find the systag for (1)
  // 3. Find workresults whose type is (2)

  // Note however that we are cheating, in that the workresults we create are of
  // type String. So we can't naively lookup the Fields using canonical foreign
  // keys. Rather, we must go through worktemplateconstraint.

  const rows = await sql<Field[]>`
    with field as (
      select
        engine1.base64_encode(convert_to('workresult:' || wr.id, 'utf8')) as id,
        s.systagtype as "type",
        wr.workresultdefaultvalue as "value"
      from entity.entityinstance
      inner join public.custag on entityinstanceoriginaluuid = custaguuid
      inner join public.worktemplateconstraint
        on custagcustomerid = worktemplateconstraintcustomerid
        and custaguuid = worktemplateconstraintconstraintid
      inner join public.workresult as wr
        on worktemplateconstraintresultid = wr.id
      inner join public.systag as s
        on wr.workresulttypeid = s.systagid
      where entityinstanceuuid = ${edge.node._id}
    )
    ${field$fragment}
  `;

  return {
    edges: rows.map(row => ({ cursor: row.id, node: row })),
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
    totalCount: rows.length,
  };
}

/**
 * @gqlField asTask
 */
export async function castEntityInstanceToTask(
  i: EntityInstance,
): Promise<Task> {
  return Task.fromTypeId(i._type, i._id);
}

/** @gqlMutationField */
export async function createCustagAsFieldTemplateValueTypeConstraint(
  ctx: Context,
  field: ID,
  name: string,
  parent: ID,
  order?: Int | null,
): Promise<Edge<EntityInstance>> {
  const { type: fieldType, id: fieldId } = decodeGlobalId(field);
  assert(fieldType === "workresult", "`field` must point to a workresult!");

  const entity_id = await sql.begin(async sql => {
    // Grab the owner id.
    const [{ owningEntity, owningLegacyCustomer }] = await sql`
      select
        entityinstance.entityinstanceuuid as "owningEntity",
        customer.customerid as "owningLegacyCustomer"
      from public.workresult
      inner join public.customer on workresultcustomerid = customerid
      inner join entity.entityinstance on customeruuid = entityinstanceoriginaluuid
      where workresult.id = ${fieldId}
    `;

    // Create the custag and entity instance.
    const [
      { create_custaguuid: custag_id, create_custagentityuuid: entity_id },
    ] = await sql`
      call entity.crud_custag_create(
        ${owningEntity},
        ${parent},
        null,
        ${order ?? null},
        ${name},
        null,
        null,
        null,
        null, 
        null, 
        null,
        null,
        null,
        auth.current_identity(${owningLegacyCustomer}, ${ctx.auth.userId})
      );
    `;

    // Create the template constraint.
    await sql`
      insert into public.worktemplateconstraint (
        worktemplateconstraintcustomerid,
        worktemplateconstraintcustomeruuid,
        worktemplateconstrainttemplateid,
        worktemplateconstraintresultid,
        worktemplateconstraintconstrainedtypeid,
        worktemplateconstraintconstraintid,
        worktemplateconstraintmodifiedby
      )
      select
        customerid,
        customeruuid,
        worktemplate.id,
        workresult.id,
        systag.systaguuid,
        custag.custaguuid,
        auth.current_identity(customerid, ${ctx.auth.userId})
      from
        public.customer,
        public.workresult,
        public.worktemplate,
        public.systag,
        public.custag
      where customerid = ${owningLegacyCustomer}
        and workresult.id = ${fieldId}
        and workresult.workresultworktemplateid = worktemplate.worktemplateid
        and custag.custaguuid = ${custag_id}
        and systag.systaguuid = (
            select entityinstanceoriginaluuid
            from entity.entityinstance
            where entityinstanceuuid = ${parent}
        )
    `;

    return entity_id;
  });

  return {
    cursor: "",
    node: {
      _id: entity_id,
      _type: "entity_instance",
      id: encodeGlobalId({ type: "entity_instance", id: entity_id }),
    },
  };
}

/** @gqlType */
export type CreateInstancePayload = {
  /** @gqlField */
  edge: Edge<EntityInstance>;
};

/** @gqlMutationField */
export async function createInstance(
  args: {
    /** @deprecated use `parent` instead */
    location?: ID | null;
    parent?: ID | null;
    template: ID;
    fields?: FieldInput[] | null;
    name?: string | null;
  },
  ctx: Context,
): Promise<CreateInstancePayload> {
  const t = new Task({ id: args.template });
  const i = await sql.begin(async sql => {
    await setCurrentIdentity(sql, ctx);
    const i = await t.instantiate(
      {
        ...args,
        parent: assertNonNull(
          args.parent ?? args.location,
          "args.parent is required",
        ),
      },
      ctx,
      sql,
    );
    return assertNonNull(i);
  });
  // NOTE: this is kinda hacky at the moment. We want everything to be an
  // EntityInstance but those guys actually map to Keller's entity model whereas
  // there is no place in the entity model at the moment for work instances. So,
  // for now, we will just fake it. Note that "entity_instance" is already taken
  // (and maps to the entity model) so we need a new type name here.
  const entityInstanceId = encodeGlobalId({ type: "instance", id: i._id });
  return {
    edge: {
      cursor: entityInstanceId,
      node: {
        _type: i._type,
        _id: i._id,
        id: entityInstanceId,
      },
    },
  };
}

//-----------------------------------------------------------------------------
//
// Entity Templates
//
// Note that currently "entity" templates are really just (and only)
// worktemplates under the hood. This will soon change.
//
//-----------------------------------------------------------------------------

export type EntityTemplateConstructorArgs = { id: ID };

/** @gqlType */
export class EntityTemplate {
  readonly _type: "template";
  readonly _id: string;
  /** @gqlField */
  readonly id: ID;

  constructor(args: EntityTemplateConstructorArgs) {
    const { type, id } = decodeGlobalId(args.id);
    this._type = assertUnderlyingType("template", type);
    this._id = id;
    this.id = args.id;
  }

  static fromTypeId(type: "template", id: string) {
    return new EntityTemplate({ id: encodeGlobalId({ type, id }) });
  }
}

/** @gqlQueryField */
export async function templates(
  args: {
    /** TEMPORARY: this should be the customer/organization uuid. */
    owner: ID;
    /**
     * Templates of the given type. This maps (currently) to worktemplatetype.
     * For example, in Runtime the following template types exist:
     * - Run
     * - Downtime
     * - Idle Time
     * Any of these are suitable for this API.
     *
     * Also see `Task.chainAgg`, as that API takes a similar parameter `overType`.
     */
    type?: string[] | null;
  },
  ctx: Context,
): Promise<Connection<EntityTemplate>> {
  const owner = decodeGlobalId(args.owner);
  assertUnderlyingType("organization", owner.type);

  const cte = args.type?.length
    ? sql`
        select
          worktemplate.id,
          worktemplate.worktemplateorder as _order
        from public.worktemplate
        inner join public.customer
          on worktemplatecustomerid = customerid
          and customeruuid = ${owner.id}
        where exists (
          select 1
          from public.worktemplatetype
          inner join public.systag on worktemplatetypesystagid = systagid
          where worktemplate.worktemplateid = worktemplatetypeworktemplateid
            and systagtype in ${sql(args.type)}
        )
      `
    : sql`
        select
          worktemplate.id,
          worktemplate.worktemplateorder as _order
        from public.worktemplate
        inner join public.customer
          on worktemplatecustomerid = customerid
          and customeruuid = ${owner.id}
      `;

  const rows = await sql<EntityTemplateConstructorArgs[]>`
    with cte as (${cte})
    select engine1.base64_encode(convert_to('template:' || cte.id, 'utf8')) as id
    from cte
    order by cte._order;
  `;

  return {
    edges: rows.map(row => ({
      cursor: row.id,
      node: new EntityTemplate(row),
    })),
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
    totalCount: rows.length,
  };
}

/** @gqlField asTask */
export function castEntityTemplateToTask(t: EntityTemplate): Task {
  if (t._type !== "template") {
    throw new GraphQLError(
      `Cannot cast Template (of type: ${t._type}) to Task`,
      {
        extensions: {
          code: "BAD_REQUEST",
        },
      },
    );
  }
  return Task.fromTypeId("worktemplate", t._id);
}
