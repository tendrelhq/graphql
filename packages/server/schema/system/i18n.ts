/**
 * A language tag in the format of a BCP 47 (RFC 5646) standard string.
 *
 * @gqlScalar
 * @specifiedBy https://www.rfc-editor.org/rfc/rfc5646.html
 */
export type Locale = string;

/**
 * Plain text content that has been (potentially) translated into different
 * languages as specified by the user's configuration.
 *
 * @gqlType
 */
export type DynamicString = {
  /** @gqlField */
  locale: Locale;
  /** @gqlField */
  value: string;
};

/** @gqlInput */
export type DynamicStringInput = {
  locale: Locale;
  value: string;
};
