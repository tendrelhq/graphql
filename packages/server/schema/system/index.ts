import z from "zod";

const GlobalId = z.string().transform(id => {
  const parts = id.split(":");
  if (parts.length < 2) {
    throw "invariant violated: invalid global identifier";
  }
  return {
    type: parts[0],
    id: parts[1],
    suffix: parts.slice(2),
  };
});

/**
 * Global Identifier. Externally it just an opaque string. Internally it may
 * have some meaning. Right now this is just the base64 encoding of the
 * object's underlying type (e.g. workinstance) and its uuid (i.e. primary
 * key).
 */
export type GlobalId = {
  type: string;
  id: string;
  suffix?: string[];
};

export function decodeGlobalId(id: unknown): GlobalId {
  return GlobalId.parse(decodeGlobalIdRaw(id));
}

export function tryDecodeGlobalId(
  id: unknown,
): { ok: true; value: GlobalId } | { ok: false; error: unknown } {
  try {
    return {
      ok: true as const,
      value: decodeGlobalId(id),
    };
  } catch (e) {
    return {
      ok: false as const,
      error: e,
    };
  }
}

export function decodeGlobalIdRaw(id: unknown): string {
  if (typeof id !== "string") {
    throw new Error(
      `invariant violated: global ids should be string but got ${typeof id}`,
    );
  }
  return Buffer.from(decodeURIComponent(id), "base64").toString();
}

export type GlobalIdInput = {
  type: string;
  id: string;
  suffix?: string | string[];
};

export function encodeGlobalId({ type, id, ...rest }: GlobalIdInput) {
  const suffix = rest.suffix
    ? typeof rest.suffix === "string"
      ? rest.suffix
      : rest.suffix.join(":")
    : "";
  return Buffer.from(
    `${type}:${id}${suffix.length ? `:${suffix}` : ""}`,
  ).toString("base64");
}

// biome-ignore lint/suspicious/noExplicitAny:
export function isGlobalId(value: any): value is GlobalId {
  if (typeof value?.type === "string" && typeof value?.id === "string") {
    return true;
  }

  return false;
}
