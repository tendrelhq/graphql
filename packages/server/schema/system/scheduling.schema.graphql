"""
A string representing a cron schedule expression
"""
scalar CronExpression

"""
Schedule implementation using cron expressions
"""
type CronSchedule {
  cron: CronExpression!
}

"""
Schedule implementation for "oneshot" events
"""
type OnceSchedule {
  once: Temporal!
}

"""
TODO
"""
union Schedule = CronSchedule | OnceSchedule

"""
TODO
"""
input ScheduleInput @oneOf {
  cron: String
  once: TemporalInput
}

"""
TODO
"""
enum ScheduleType {
  CronSchedule
  OnceSchedule
}
