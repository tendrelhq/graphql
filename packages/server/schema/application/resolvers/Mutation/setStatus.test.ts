import { describe, expect, test } from "bun:test";
import { schema } from "@/schema/final";
import { encodeGlobalId } from "@/schema/system";
import { createTestContext, execute, testGlobalId } from "@/test/prelude";
import { TestSetStatusDocument } from "./setStatus.test.generated";

const ctx = await createTestContext();

describe.skip("setStatus", () => {
  test("workinstance", async () => {
    const result = await execute(ctx, schema, TestSetStatusDocument, {
      entity: encodeGlobalId({
        type: "workinstance",
        id: "work-instance_35774847-ba19-4f23-b71e-e4cb7760e415",
      }),
      input: {
        // To test with delta:
        // 1. run as is; will pass
        // 2. change to inProgress, run again; will fail with delta AND status diff
        // 3. change back to open, run again; will fail with delta diff
        // 4. run again; will pass
        open: {
          at: {
            instant: Date.now().toString(),
          },
        },
      },
    });

    expect(result).toMatchSnapshot();
  });

  test("workresultinstance", async () => {
    const result = await execute(ctx, schema, TestSetStatusDocument, {
      parent: encodeGlobalId({
        type: "workinstance",
        id: "work-instance_35774847-ba19-4f23-b71e-e4cb7760e415",
      }),
      entity: encodeGlobalId({
        type: "workresultinstance",
        id: "work-instance_35774847-ba19-4f23-b71e-e4cb7760e415",
        suffix: "work-result_a3f48541-bf6c-4739-94ee-985ab629e1e3",
      }),
      input: {
        // Follow the same procedure described above to test with delta.
        closed: {
          at: {
            instant: Date.now().toString(),
          },
        },
      },
    });

    expect(result).toMatchSnapshot();
  });

  test("invalid status change", async () => {
    const result = await execute(ctx, schema, TestSetStatusDocument, {
      parent: encodeGlobalId({
        type: "workinstance",
        id: "work-instance_93c61cb5-e5ec-43e1-8777-d9f6e930e6b0",
      }),
      entity: encodeGlobalId({
        type: "workresultinstance",
        id: "work-instance_93c61cb5-e5ec-43e1-8777-d9f6e930e6b0",
        suffix: "work-result_fa536e61-2e9f-480c-b815-5cb1ec0a0f79",
      }),
      input: {
        inProgress: {
          at: {
            instant: Date.now().toString(),
          },
        },
      },
    });

    expect(result).toMatchSnapshot();
  });

  test("entity cannot have its status changed", async () => {
    const result = await execute(ctx, schema, TestSetStatusDocument, {
      entity: testGlobalId(),
      input: {
        open: {
          at: {
            instant: Date.now().toString(),
          },
        },
      },
    });
    expect(result).toMatchSnapshot();
  });
});
