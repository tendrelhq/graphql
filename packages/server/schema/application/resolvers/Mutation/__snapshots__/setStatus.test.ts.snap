// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`setStatus workinstance 1`] = `
{
  "data": {
    "setStatus": {
      "__typename": "SetChecklistStatusPayload",
      "delta": 0,
      "edge": {
        "__typename": "ChecklistEdge",
        "node": {
          "__typename": "Checklist",
          "assignees": {
            "__typename": "AssigneeConnection",
            "edges": [
              {
                "__typename": "AssigneeEdge",
                "node": {
                  "__typename": "Assignee",
                  "assignedTo": {
                    "__typename": "Worker",
                    "displayName": "Will Ruggiano",
                  },
                },
              },
            ],
          },
          "status": {
            "__typename": "ChecklistOpen",
          },
        },
      },
    },
  },
}
`;

exports[`setStatus workresultinstance 1`] = `
{
  "data": {
    "setStatus": {
      "__typename": "SetChecklistItemStatusPayload",
      "delta": 0,
      "edge": {
        "__typename": "ChecklistItemEdge",
        "node": {
          "__typename": "ChecklistResult",
          "status": {
            "__typename": "ChecklistClosed",
          },
        },
      },
    },
  },
}
`;

exports[`setStatus invalid status change 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: Invalid status change],
  ],
}
`;

exports[`setStatus entity cannot have its status changed 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: Entity cannot have its status changed],
  ],
}
`;
